R/main.R

```diff
@@ -114,68 +134,97 @@ Options:
 
 Parameters, if given, must be passed through in key=value pairs"
 
-
-main_do_run <- function(x) {
-  ## TODO: Get some classed errors though here and then write out
-  ## information about whether or not things worked and why they
-  ## didn't.  Possible issues (in order)
-  ##
-  ## * orderly report not found
-  ## * error while preparing (e.g., package not found)
-  ## * error while running report
-  ## * error checking artefacts
-  config <- orderly_config_get(x$options$root, TRUE)
-  name <- x$options$name
-  commit <- !x$options$no_commit
-  instance <- x$options$instance
-  parameters <- x$options$parameters
-  id_file <- x$options$id_file
-  parameters <- x$options$parameters
-  print_log <- x$options$print_log
-  ref <- x$options$ref
-  fetch <- x$options$fetch
-  pull <- x$options$pull
-  message <- x$options$message
-
-  main_run <- function() {
-    if (pull) {
-      if (is.null(ref)) {
-        git_pull(config$root)
-      } else {
-        orderly_cli_error(
-          "Can't use --pull with --ref; perhaps you meant --fetch ?")
+do_run <- function(run_func) {
+  function(x) {
+    ## TODO: Get some classed errors though here and then write out
+    ## information about whether or not things worked and why they
+    ## didn't.  Possible issues (in order)
+    ##
+    ## * orderly report not found
+    ## * error while preparing (e.g., package not found)
+    ## * error while running report
+    ## * error checking artefacts
+    config <- orderly_config_get(x$options$root, TRUE)
+    name <- x$options$name
+    commit <- !x$options$no_commit
+    instance <- x$options$instance
+    parameters <- x$options$parameters
+    id_file <- x$options$id_file
+    parameters <- x$options$parameters
+    print_log <- x$options$print_log
+    ref <- x$options$ref
+    fetch <- x$options$fetch
+    pull <- x$options$pull
+    message <- x$options$message
+
+    main_batch <- function() {
+      if (pull) {
+        if (is.null(ref)) {
+          git_pull(config$root)
+        } else {
+          orderly_cli_error(
+            "Can't use --pull with --ref; perhaps you meant --fetch ?")
+        }
       }
+      ids <- run_func(name, parameters, root = config, id_file = id_file,
+                      instance = instance, ref = ref, fetch = fetch,
+                      message = message)
+      if (commit) {
+        lapply(ids, orderly_commit, name, config)
+      }
+      ids
+    }
+
+    if (print_log) {
+      sink(stderr(), type = "output")
+      on.exit(sink(NULL, type = "output"))
+      ids <- main_batch()
+    } else {
+      log <- tempfile()
+      ## we should run this with try() so that we can capture logs there
+      ids <- capture_log(main_batch(), log)
+      dest <- (if (commit) path_archive else path_draft)(config$root)
+      ## TODO: Here all orderly logs get the history for all reports in the
```

Comment:
What do you think about this comment? Do we want the logs for all reports in the batch to be saved for every report within the batch?

Reply:
yeah, logs need to go into their individual directories